From abf2fe713739dcca6344658b216b35d90a3bb758 Mon Sep 17 00:00:00 2001
From: Noam Camus <ncamus@habana.ai>
Date: Sat, 25 Apr 2020 07:52:52 +0300
Subject: [PATCH] [SW-0] uart: add new command module

It will be used to parse serial pressed keys
Keys that will signal some special behavior
Like Esc or CtrlC characters

Change-Id: I0b2a36b98e50c2887913797adbe6c04ae094c04c
---

diff --git a/app/preboot.c b/app/preboot.c
index eed483d..e2d935a 100644
--- a/app/preboot.c
+++ b/app/preboot.c
@@ -56,6 +56,16 @@
 		.teardown = module_lkd_disable,
 		.run_noreturn = module_lkd_run_noret,
 	},
+	{	/* UART module */
+		.setup = noop,
+		.check = module_uart_check,
+		.disable = noop,
+		.run = noop,
+		.enable = noop,
+		.run_fast = module_uart_run,
+		.teardown = noop,
+		.run_noreturn = noop,
+	},
 };
 
 static void preboot_handle_cmds(void)
diff --git a/common/cmd_module.h b/common/cmd_module.h
index 7327c60..c2d57b9 100644
--- a/common/cmd_module.h
+++ b/common/cmd_module.h
@@ -78,4 +78,8 @@
 void module_lkd_run(void);
 void module_lkd_run_noret(void);
 
+/* UART */
+enum cmd_check_res module_uart_check(void);
+void module_uart_run(void);
+
 #endif /* _HL_CMD_MODULE_H_ */
diff --git a/common/module_uart.c b/common/module_uart.c
new file mode 100644
index 0000000..16f3259
--- /dev/null
+++ b/common/module_uart.c
@@ -0,0 +1,32 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright 2016-2020 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+#include <preboot.h>
+
+#include "cmd_module.h"
+#include "uart.h"
+
+enum cmd_check_res module_uart_check(void)
+{
+	if (!preboot_uart_tstc())
+		return NOT_FOUND;
+
+	preboot_puts("Keyboard press received\n");
+
+	return FOUND_FAST_RETURN;
+}
+
+__attribute__((weak)) void hl_parse_uart(int val) {}
+
+void module_uart_run(void)
+{
+	int val = preboot_uart_getc();
+
+	if (val < 0)
+		return;
+
+	hl_parse_uart(val);
+}
